dev/benchmarking/ddl/1_tables.sql

```diff
@@ -0,0 +1,291 @@
+  IS 'The output of `sysctl -n hw.cpufrequency_max`.';
+COMMENT ON COLUMN public.cpu.cpu_frequency_min_Hz
+  IS 'The output of `sysctl -n hw.cpufrequency_min`.';
+COMMENT ON COLUMN public.cpu.cpu_L1d_cache_bytes
+  IS 'The output of `sysctl -n hw.l1dcachesize`.';
+COMMENT ON COLUMN public.cpu.cpu_L1i_cache_bytes
+  IS 'The output of `sysctl -n hw.l1icachesize`.';
+COMMENT ON COLUMN public.cpu.cpu_L2_cache_bytes
+  IS 'The output of `sysctl -n hw.l2cachesize`.';
+COMMENT ON COLUMN public.cpu.cpu_L3_cache_bytes
+  IS 'The output of `sysctl -n hw.l3cachesize`.';
+COMMENT ON COLUMN public.cpu.cpu_RAM_bytes
+  IS 'The output of `sysctl -n machdep.cpu.cache.size`.';
+
+-- GPU
+CREATE TABLE IF NOT EXISTS public.gpu
+(
+  id SERIAL PRIMARY KEY
+  , gpu_model_name citext UNIQUE NOT NULL DEFAULT ''
+  , gpu_core_count integer
+  , gpu_RAM_bytes bigint
+);
+COMMENT ON TABLE public.gpu IS 'GPU specifications.';
+
+-- OS
+CREATE TABLE IF NOT EXISTS public.os
+(
+  id SERIAL PRIMARY KEY
+  , os_name citext NOT NULL
+  , architecture_name citext NOT NULL
+  , kernel_name citext NOT NULL DEFAULT ''
+);
+-- @name os. forces retention of an `s` in the Graphile GraphQL api.
+COMMENT ON TABLE public.os
+  IS E'@name os.\nOperating system name and kernel (version).';
+COMMENT ON COLUMN public.os.os_name
+  IS 'Operating system name. For example, OSX, Ubuntu, Windows`.';
+COMMENT ON COLUMN public.os.architecture_name
+  IS 'Operating system architecture; the output of `uname -m`.';
+COMMENT ON COLUMN public.os.kernel_name
+  IS 'Operating system kernel, or NULL. '
+     'On Linux/OSX, the output of `uname -r`. '
+     'On Windows, the output of `ver`.';
+
+-- PROJECT
+CREATE TABLE IF NOT EXISTS public.project
+(
+  id SERIAL PRIMARY KEY
+  , project_name citext NOT NULL
+  , project_url text NOT NULL
+  , repo_url text NOT NULL
+  , last_changed timestamp (0) without time zone NOT NULL DEFAULT now()
+);
+COMMENT ON TABLE public.project
+  IS 'Project name and relevant URLs.';
+COMMENT ON COLUMN public.project.project_url
+  IS 'Homepage URL.';
+COMMENT ON COLUMN public.project.repo_url
+  IS 'Git repo URL to link stored commit hashes to code in a webpage.';
+COMMENT ON COLUMN public.project.last_changed
+  IS 'New project details are added with a new timestamp. '
+     'The project details with the newest timestamp will be used.';
+
+-- BENCHMARK_LANGUAGE
+CREATE TABLE IF NOT EXISTS public.benchmark_language
+(
+  id SERIAL PRIMARY KEY
+  , benchmark_language citext NOT NULL UNIQUE
+);
+COMMENT ON TABLE public.benchmark_language
+  IS 'The language the benchmark was written in (and presumably for).';
+COMMENT ON COLUMN public.benchmark_language.benchmark_language
+  IS 'The benchmark language. For example: Python';
+
+-- LANGUAGE_DEPENDENCIES
+CREATE TABLE IF NOT EXISTS public.language_dependencies
+(
+  id SERIAL
+  , benchmark_language_id integer NOT NULL
+  , dependencies jsonb NOT NULL DEFAULT '{}'::jsonb
+  , PRIMARY KEY (id, benchmark_language_id)
```

Comment:
why use the pair here?

Reply:
This was to handle carrying the benchmark language through to the environment definition. If it's the primary key, then when I foreign-key it into `environment`, and force the same `benchmark_language_id` for both the `language_version` and `language_dependencies`, I am able to be sure that the `environment` will only ever be associated with one language, even though it depends on two tables.

I think this solution is inelegant but was at a loss for something better, and would gratefully accept other design ideas. (For example, maybe just skipping the `language_version` and `language_dependencies` tables and  adding uniqueness constraints in `environment`?) For the record, [here (png)](https://issues.apache.org/jira/secure/attachment/12957979/benchmark-data-model.png) is a mostly up-to-date (without the new `run_timestamp` column) diagram of the data model in [the JIRA](https://issues.apache.org/jira/browse/ARROW-4313)). Depending on the `benchmark_language` table makes everything downstream more complicated than I like.